components/ide/jetbrains/launcher/log_analyzer.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2024 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License.AGPL.txt in the project root for license information.
+
+package main
+
+import (
+	"context"
+	"os"
+	"path"
+	"regexp"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"github.com/gitpod-io/gitpod/jetbrains/launcher/pkg/metrics"
+	"github.com/nxadm/tail"
+)
+
+type LogAnalyzer interface {
+	Analyze(ctx context.Context) error
+}
+
+const (
+	logDir = "/tmp/ide-desktop-log"
+)
+
+func init() {
+	if err := os.MkdirAll(logDir, 0755); err != nil {
+		log.WithError(err).Error("failed to create logDir")
+	}
+}
+
+type LineMatchRule struct {
+	Name           string         // rule name
+	Pattern        *regexp.Regexp // regex to match line
+	LogFile        string         // print matched line to log file
+	matchedHandler func()
+
+	matched bool
+}
+
+// IdeaLogFileAnalyzer watches the idea.log file and does diagnostic on the output
+type IdeaLogFileAnalyzer struct {
+	launchCtx *LaunchContext
+	path      string
+	rules     []*LineMatchRule
+}
+
+var _ LogAnalyzer = &IdeaLogFileAnalyzer{}
+
+func NewIdeaLogAnalyzer(launchCtx *LaunchContext, logPath string) *IdeaLogFileAnalyzer {
+	ide := "unknown"
+	if launchCtx != nil {
+		ide = launchCtx.alias
+	}
+	logDir := path.Dir(logPath)
+	if err := os.MkdirAll(logDir, 0755); err != nil {
+		// if error and the directory does not exist, watcher will close chan directly
+		// so no need to return error here
+		log.WithError(err).Error("failed to create log file's directory")
+	}
+
+	l := &IdeaLogFileAnalyzer{
+		path:      logPath,
+		launchCtx: launchCtx,
+		rules: []*LineMatchRule{
+			{Name: "plugin started", Pattern: regexp.MustCompile(`Gitpod gateway link`), LogFile: "jb-backend-started.log", matchedHandler: func() {
```

Comment:
@mustard-mh did you test JB backend restart? idea.log is appended? would double metrics each time or only pick up new?

Reply:
We probably should find last start in logs and proceed from this point or something like that?

Reply:
hm, it is also tricky, because JB backend can be restarted by different processes:
- supervisor
- launcher
- jetbrains shell script

We probably cannot correlate plugins to jb starts if we measure it only from supervisor?

Reply:
JB backend restart case
- supervisor: (i.e. workspace restart) yes this will happen
- launcher: we don't do retry or restart on launcher, so it never happen
- jetbrains shell script: we can't control it, even we do, we don't know supervisor_ide is restarted. so we should ignore this case

Base on ☝️ , we could rename idea.log when we're going to start launcher, WDYT? @akosyakov